### PR TITLE
Allow array unpacking in ArrayDeclaration multiline Sniff

### LINKED ISSUE #3557: Unpacking associative arrays in PHP 8.1 produces `Squiz.Arrays.ArrayDeclaration.NoKeySpecified` respectively `Squiz.Arrays.ArrayDeclaration.KeySpecified`
**Describe the bug**
Since PHP 8.1 unpacking of associative arrays is supported.
Unpacking an associative array into another results in one of the errors `Squiz.Arrays.ArrayDeclaration.NoKeySpecified` or `Squiz.Arrays.ArrayDeclaration.KeySpecified` (depending on the position).

**Code sample**
```php
<?php

$arr = [
        'a' => 1,
        'b' => 2,
       ];

$merged1 = [
            'c' => 3,
            ...$arr,
           ];

$merged2 = [
            ...$arr,
            'c' => 3,
           ];
```

**Custom ruleset**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<ruleset>
    <rule ref="Squiz.Arrays.ArrayDeclaration"/>
</ruleset>
```

**To reproduce**
```
-------------------------------------------------------------------------------------------------------------------------
FOUND 2 ERRORS AFFECTING 2 LINES
-------------------------------------------------------------------------------------------------------------------------
 10 | ERROR | No key specified for array entry; first entry specifies key
    |       | (Squiz.Arrays.ArrayDeclaration.NoKeySpecified)
 15 | ERROR | Key specified for array entry; first entry has no key (Squiz.Arrays.ArrayDeclaration.KeySpecified)
-------------------------------------------------------------------------------------------------------------------------
```

**Expected behavior**
This should not result in an error:
* ideally PHP_CodeSniffer can check if the unpacked array is associative
* if thats not possible, array unpacking should be ignored by this rule

**Versions:**
 - OS: Windows 10
 - PHP: 8.1
 - PHPCS: 3.6.2
 - Standard: Squiz

### PR DESCRIPTION
**Pull request by edorian**
See #3557

## Description

`Squiz.Arrays.ArrayDeclaration.NoKeySpecified` doesn't accept the following snippet:

```php
$x = [
  'foo' => 'bar',
  ...$baz,
];
```

It triggers an error as phpcs doesn't detect that the second entry is array unpacking.

```
Key specified for array entry; first entry has no key (Squiz.Arrays.ArrayDeclaration.KeySpecified)
``` 

The following (single line) array validates:

```php
$x = ['foo' => 'bar',  ...$baz];
```

### Suggested changelog entry

`Squiz.Arrays.ArrayDeclaration.KeySpecified`: Allow array unpacking (`...$array`) in multi-line key-value array definitions.

### Related issues/external references

Fixes #3557

## Types of changes

- [x] Bug fix _(non-breaking change which fixes an issue)_
- [ ] New feature _(non-breaking change which adds functionality)_
- [ ] Breaking change _(fix or feature that would cause existing functionality to change)_
    - [ ] This change is only breaking for integrators, not for external standards or end-users.
- [ ] Documentation improvement

## PR checklist

- [x] I have checked there is no other PR open for the same change.
- [x] I have read the [Contribution Guidelines](.github/CONTRIBUTING.md).
- [x] I grant the project the right to include and distribute the code under the BSD-3-Clause license (and I have the right to grant these rights).
- [x] I have added tests to cover my changes.
  - I did not. I'd be happy to, if you give me a pointer on how I can do this without breaking earlier PHP versions, given it's new syntax.
- [x] I have verified that the code complies with the projects coding standards.
- [ ] [Required for new sniffs] I have added XML documentation for the sniff.
- [ ] [Required for new files] I have added any new files to the `package.xml` file.

## Notes:

The one test failure for PHP 8.2 seems unrelated as it was already present in https://github.com/squizlabs/PHP_CodeSniffer/commit/2e59060b93b3720ff3aab2b57efda0c3ada069b0

### CODE REVIEW

**Comment by jrfnl** [comment]
> I did not. I'd be happy to, if you give me a pointer on how I can do this without breaking earlier PHP versions, given it's new syntax.

@edorian Please do add some new tests. The fact that it is new syntax is irrelevant as the PHPCS tokenizer has backported the `T_ELLIPSIS` token to PHP 5.4, so the test code should tokenize the same in all PHP versions.

(if that is not the case, we have another issue, but better to know about it than not)

**Comment by edorian** [comment]
[x] Test

Without my fix:

```
PHP_CodeSniffer\Standards\Squiz\Tests\Arrays\ArrayDeclarationUnitTest::testSniff
[LINE 475] Expected 0 error(s) in ArrayDeclarationUnitTest.2.inc but found 1 error(s). The error(s) found were:
 -> No key specified for array entry; first entry specifies key (Squiz.Arrays.ArrayDeclaration.NoKeySpecified)
```

With my fix :heavy_check_mark:

File: src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php [hunk 1]
```diff
@@ -430,9 +430,22 @@ public function processMultiLineArray($phpcsFile, $stackPtr, $arrayStart, $array
...
+
+                    /*
+                        Allow array unpacking
+
+                        $x = [
+                          'foo' => 'bar',
+                          ...$baz,
+                        ];
+                    */
+
```
Comment (jrfnl) [inline src/Standards/Squiz/Sniffs/Arrays/ArrayDeclarationSniff.php:443]:
```suggestion
                    // Allow for PHP 7.4 array unpacking with an array declaration.
```

I don't think the code sample is needed here.

File: src/Standards/Squiz/Tests/Arrays/ArrayDeclarationUnitTest.2.inc [hunk 2]
```diff
@@ -468,3 +468,9 @@ yield [
 $a = [
       'a' =>
      ];
+
+// Ensure absence of NoKeySpecified error for array unpacking
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Arrays/ArrayDeclarationUnitTest.2.inc:472]:
```suggestion
// Ensure absence of NoKeySpecified error for PHP 7.4+ array unpacking.
```

**Review by jrfnl — CHANGES_REQUESTED** [review]
@edorian Hi Volker, thanks for this PR.

I can confirm the current fix is correct and working.

I've left some minor comments inline, but nothing blocking.

One thing I would like to ask - could you add the test to both the test case files ?

The `ArrayDeclarationUnitTest.1.inc` test case file is supposed to have long array syntax, while the `ArrayDeclarationUnitTest.2.2inc` test case file is supposed to have those same tests using short array syntax.

It does look like the files have diverged a little over time, but I do think it's good to have the test(s) in both files.

And looking at the test, I also wonder if there should also be a test with the array unpacking at the start of the array declaration and one where the array unpacking happens in the "middle" of an array ?

As things are, this PR addresses just one part of issue #3557 (`NoKeySpecified`), while the other part (`KeySpecified`) is not yet fixed.

**Comment by edorian** [comment]
@jrfnl I've implemented your suggestions, fixed the KeySpecified error and added test cases.

I've done so by basically saying "if it's array unpacking line don't decide if it's a 'list' or a 'dict' shaped array".

Doesn't look very clean looking at the diff but it passes all tests I could come up with :)

File: src/Standards/Squiz/Tests/Arrays/ArrayDeclarationUnitTest.1.inc [hunk 3]
```diff
@@ -486,3 +486,20 @@ $foo = [
 $a = array(
       'a' =>
      );
+
+// Ensure absence of NoKeySpecified error for PHP 7.4+ array unpacking.
```
Comment (jrfnl) [inline src/Standards/Squiz/Tests/Arrays/ArrayDeclarationUnitTest.1.inc:490]: This comment only refers to the one error code, while both are now handled. Maybe something like this ?
```suggestion
// Safeguard correct errors for key/no key when PHP 7.4+ array unpacking is encountered.
```

**Review by jrfnl — COMMENTED** [review]
@edorian Thanks for this update. The fixes look good.

I'm just thinking that there should probably be some additional tests still to safeguard that when an array entry is ignored due to unpacking, the key/no key for literal array entries is still correct (which it is with your fixes, but safeguarding that is the idea).

What about adding these additional tests ?

```php
$x = [
      ...$a,
      'foo' => 'bar', // OK.
      'bar', // Error about no key (based on second entry).
     ];

$x = [
      ...$a,
      'bar', // OK.
      'foo' => 'bar', // Error about key (based on second entry).
     ];

$x = [
      'foo' => 'bar',
      ...$a,
      'baz' => 'bar',
      'bar', // Error about no key (based on first entry).
     ];

$x = [
      'bar',
      ...$a,
      'bar',
      'baz' => 'bar', // Error about key (based on first entry).
     ];
```

P.S.: thanks for taking on this issue, this is one of the sniffs which is unwieldy to adjust in most cases. Also see #582.

**Comment by edorian** [comment]
Added your suggested tests and tried to align the two test files, but the diff of that grew to big for this PR. 

Happy to do that in a follow up when this is merged :) 

@jrfnl Thank you for the fast response an the clear directions!

File: src/Standards/Squiz/Tests/Arrays/ArrayDeclarationUnitTest.2.inc [hunk 4]
```diff
@@ -464,7 +464,57 @@ yield [
        static fn () : string => '',
       ];
 
+$foo = [
+        'foo' => match ($anything) {
+            'foo' => 'bar',
+            default => null,
+        },
+       ];
+
```
Comment (edorian) [inline src/Standards/Squiz/Tests/Arrays/ArrayDeclarationUnitTest.2.inc:473]: This is aligning 1.inc and 2.inc

**Comment by jrfnl** [comment]
@edorian Uh oh.. the tests are failing now. Could it be that the test case comment was only applied to the one test case file, not to all four ?

**Review by jrfnl — APPROVED** [review]
> Didn't commit/push all the changes.

@edorian Happens to the best of us.

All good now! Verified the test fail without the fix and pass with the correct expected notices with the fixes. ✅

**Comment by jrfnl** [comment]
FYI: this fix is included in today's [PHP_CodeSniffer 3.8.0 release](https://github.com/PHPCSStandards/PHP_CodeSniffer/releases/tag/3.8.0).

As per #3932, development on PHP_CodeSniffer will continue in the [PHPCSStandards/PHP_CodeSniffer](https://github.com/PHPCSStandards/PHP_CodeSniffer) repository. If you want to stay informed, you may want to start "watching" that repo (or watching releases from that repo).

**Pull request merged**